### ISSUE TITLE
NTXentLoss

### ISSUE THREAD

**Issue by su-0618-sudo** [KevinMusgrave/pytorch-metric-learning#463]
I'm a beginner. The question may be funny, but please include it. I want to know if ntxentloss uses the Simclr method. For this self-supervised method, what is the purpose of entering labels? Whether to determine which are positive samples and which are negative samples？ So as not to widen the distance of similar samples. And whether MoCo method has no class that can be called directly. Does it mean that I can refer to mococicar10 ipynb this notebook. If I want to execute this notebook, I need GPU? I only have CPU now.hhhh

**Comment by KevinMusgrave** [comment]
> I want to know if ntxentloss uses the Simclr method. 

It's the other way around, SimCLR uses NTXent also known as InfoNCE.

> For this self-supervised method, what is the purpose of entering labels? Whether to determine which are positive samples and which are negative samples？So as not to widen the distance of similar samples.

Correct, in this library you provide labels to indicate which samples belong together and which ones should be far apart. In version 2.0, there will be a ```SelfSupervisedLossWrapper``` so that you don't have to create the labels yourself (see issue https://github.com/KevinMusgrave/pytorch-metric-learning/issues/411 for details)

> And whether MoCo method has no class that can be called directly. Does it mean that I can refer to mococicar10 ipynb this notebook. 

Correct, this library does not yet provide a simple class for SimCLR or MoCo. It only provides the loss function. 

> If I want to execute this notebook, I need GPU? I only have CPU now.hhhh

If you're running it in google colab, you can go to the Runtime menu, and click on "Change runtime type" to make it run on CPU. It will be very slow though.

Here are some github repos you might find useful:
MoCo: https://github.com/facebookresearch/moco
SimCLR: https://github.com/sthalles/SimCLR

**Comment by su-0618-sudo** [comment]
Thank you very much for giving such detailed guidance to a beginner. I gain a lot. Now I find that my understanding of some concepts is biased. I always thought that when I used this class, it encapsulated the simclr method. I think its internal logic is the following formula in SimCLR. But its input label is to avoid stretching the distance of similar samples. It seems that I misunderstood.

![question](https://user-images.githubusercontent.com/65110171/163772377-1086730d-8167-4d9b-be1c-1271f21463df.png)

According to your explanation, this is neither simclr nor Moco? Is that what you mean? Can this be understood as supervised comparative loss? The principle behind this function is to construct all samples in the same class into positive sample pairs, and then widen the distance from negative samples? So it seems that using this loss is more efficient than simclr method, isn't it? Because according to my understanding, simclr may extend the distance of similar samples. Thank you again for your reply.

**Comment by KevinMusgrave** [comment]
SimCLR and MoCo both use NTXent as the loss function. But you can use NTXent in other situations, like in supervised learning rather than self-supervised learning. So NTXent by itself has nothing to do with SimCLR and MoCo. It is simply a loss function that is used by those self-supervised methods.
